### PR TITLE
[TEP-0089] - Phase 1 Signed TaskRun Results

### PR DESCRIPTION
**Pull request by pxp928**
Signed-off-by: pxp928 <parth.psu@gmail.com>

# Changes

Authors - @pxp928 and @lumjjb 

In association with [TEP-0089: Non-falsifiable provenance support](https://github.com/tektoncd/community/blob/main/teps/0089-nonfalsifiable-provenance-support.md)

This PR is the implementation of Phase 1 of the [TEP-0089: Non-falsifiable provenance support](https://github.com/tektoncd/community/blob/main/teps/0089-nonfalsifiable-provenance-support.md)

Phase 2 of the PR. Thats build on top of this to add Signed TaskRun Status can be found here - #4828

Phase 1

* Add support for Signed Results with SPIRE (this will primarily involve modifications to the entrypointer image)
* Add support for tekton-pipelines-controller verifying Signed Results

Taking the work that @dlorenc started a while back and adding improvements to get the spire integrated with Tekton pipeline.

The integration of spire k8s workload registrar automatically allows for tekton created pods to be registered into the spire-server.

Currently spire-server and spire-agent needs to be running in your cluster in order for tekton pipelines to integrate. Please follow the Spire documentation that is part of this PR to set up for local testing.

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [x] [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) included if any changes are user facing
- [x] [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [x] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [x] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including
  functionality, content, code)
- [x] Release notes block below has been filled in or deleted (only if no user facing changes)

# Release Notes

- Added Spiffe-CSI driver to the controller and allow for spire workload API communication over CSI driver (not using hostpath)
- Added in [k8s-workload-registrar](https://github.com/spiffe/spire/tree/main/support/k8s/k8s-workload-registrar) into the tekton pipeline controller. This will allow pods to be auto registered into the spire server based on pod annotations. 
- Added fetch SVID based on the pod running the TaskRun and use the private key to sign payload and attach the corresponding SVID to the termination messages
- Add in a RESULT_MANIFEST to register all outputs of the TaskRun
- Tekton Pipelines verifies the Results against the SPIRE SVID and Trust Bundle and sets the SignedResultsVerified condition to True

Please provide feedback and improvements!

### CODE REVIEW

**Comment by pxp928** [comment]
@pritidesai Can we get another ok to test?

File: pkg/pod/pod.go [hunk 1]
```diff
@@ -279,11 +288,37 @@ func (b *Builder) Build(ctx context.Context, taskRun *v1beta1.TaskRun, taskSpec
...
+				CSI: &corev1.CSIVolumeSource{
+					Driver: "csi.spiffe.io",
+				},
+			},
+		})
+
+		for i := range stepContainers {
+			c := &stepContainers[i]
+			c.VolumeMounts = append(c.VolumeMounts, corev1.VolumeMount{
+				Name:      "spiffe-workload-api",
```
Comment (dlorenc) [inline pkg/pod/pod.go:309]: Nit: should we make these constants?

File: pkg/pod/status.go [hunk 2]
```diff
@@ -140,7 +147,30 @@ func MakeTaskRunStatus(logger *zap.SugaredLogger, tr v1beta1.TaskRun, pod *corev
...
+			logger.Errorf("failed to verify signed results with spire: %w", err)
+			markStatusSignedResultsFailure(trs, err.Error())
+		} else {
+			logger.Info("successfully validated signed results with spire")
+			markStatusSignedResultsVerified(trs)
+		}
+	}
+
+	// If no results and no results requested, set verified unless results were specified as part of task spec
+	if len(filteredResults) == 0 && (tr.Status.TaskSpec == nil || len(tr.Status.TaskSpec.Results) == 0) {
```
Comment (dlorenc) [inline pkg/pod/status.go:166]: Hmm, this feels a bit dangerous, would it be safer to return false here if there were no results to verify?

**Review by dlorenc — COMMENTED** [review]
Overall looks great! A few small nits.

File: pkg/pod/status.go (hunk 2 above)
Comment (lumjjb) [reply pkg/pod/status.go:166]: (This is assuming the phase 2 PR which assumes integrity of the Status.TaskSpec.

File: pkg/pod/pod.go (hunk 1 above)
Comment (pxp928) [reply pkg/pod/pod.go:309]: Can do.

File: pkg/pod/status.go (hunk 2 above)
Comment (pxp928) [reply pkg/pod/status.go:166]: Agree with @lumjjb. As Phase 2 of TEP 0089 would see removal of results by a malicious actor, that would fail the overall TaskRun Status verification. If results are expected but removed, we do have another check for the `RESULT_MANIFEST` that will check that the results requested are returned. If that does not match, it will fail the verification.

File: cmd/imagedigestexporter/main.go [hunk 3]
```diff
@@ -31,6 +34,8 @@ import (
 var (
 	images                 = flag.String("images", "", "List of images resources built by task in json format")
 	terminationMessagePath = flag.String("terminationMessagePath", "/tekton/termination", "Location of file containing termination message")
+	enableSpire            = flag.Bool("enable_spire", false, "If specified by configmap, this enables spire signing and verification")
+	socketPath             = flag.String("spire-socket-path", "/spiffe-workload-api/spire-agent.sock", "Experimental: The SPIRE agent socket for SPIFFE workload API.")
```
Comment (priyawadhwa) [inline cmd/imagedigestexporter/main.go:38]: same nit as above

File: docs/spire.md [hunk 4]
```diff
@@ -0,0 +1,284 @@
+<!--
+---
+linkTitle: "SPIRE"
+weight: 1660
+---
+-->
+# SPIRE TaskRun result attestations
+
+The SPIRE TaskRun result attestations feature provides the non-falsifiable provenance to the build processes that run in the pipeline. They ensure that the results of the tekton pipeline executions originate from the build workloads themselves and that they have not been tampered.
```
Comment (priyawadhwa) [inline docs/spire.md:9]: nit: the first sentence makes it seem like result attestations alone deliver non-falsifiable provenance. can you clarify that this is one part of achieving non-falsifiable provenance, or that it guarantees non-falsifiable results specifically?

File: pkg/spire/entrypointer.go [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+func NewSpireEntrypointerAPIClient(c spireconfig.SpireConfig) EntrypointerAPIClient {
+	return &spireEntrypointerAPIClient{
+		config: c,
+	}
+}
+
+func (w *spireEntrypointerAPIClient) Close() {
+	err := w.client.Close()
+	if err != nil {
+		// Log error
```
Comment (priyawadhwa) [inline pkg/spire/entrypointer.go:75]: log error?

**Review by priyawadhwa — COMMENTED** [review]
This is looking really good! Left a few nits and questions.

File: cmd/entrypoint/main.go [hunk 6]
```diff
@@ -49,6 +51,8 @@ var (
 	onError             = flag.String("on_error", "", "Set to \"continue\" to ignore an error and continue when a container terminates with a non-zero exit code."+
 		" Set to \"stopAndFail\" to declare a failure with a step error and stop executing the rest of the steps.")
 	stepMetadataDir = flag.String("step_metadata_dir", "", "If specified, create directory to store the step metadata e.g. /tekton/steps/<step-name>/")
+	enableSpire     = flag.Bool("enable_spire", false, "If specified by configmap, this enables spire signing and verification")
+	socketPath      = flag.String("spire-socket-path", "/spiffe-workload-api/spire-agent.sock", "Experimental: The SPIRE agent socket for SPIFFE workload API.")
```
[the remark this answers is not in the document]
Comment (pxp928) [reply cmd/entrypoint/main.go:55]: Fixed.

File: cmd/imagedigestexporter/main.go (hunk 3 above)
Comment (pxp928) [reply cmd/imagedigestexporter/main.go:38]: fixed.

File: docs/spire.md (hunk 4 above)
Comment (pxp928) [reply docs/spire.md:9]: Good catch. Changed.

File: docs/spire.md [hunk 7]
```diff
@@ -0,0 +1,284 @@
...
+---
+linkTitle: "SPIRE"
+weight: 1660
+---
+-->
+# SPIRE TaskRun result attestations
+
+The SPIRE TaskRun result attestations feature provides the non-falsifiable provenance to the build processes that run in the pipeline. They ensure that the results of the tekton pipeline executions originate from the build workloads themselves and that they have not been tampered. 
+
+When SPIRE TaskRun result attestations is enabled, all TaskRuns will produce a signature alongside its results, which can then be used to validate its provenance. For example, a TaskRun result would look like this:
```
[the remark this answers is not in the document]
Comment (pxp928) [reply docs/spire.md:11]: Added.

File: test/taskrun_test.go [hunk 8]
```diff
@@ -136,16 +156,27 @@ spec:
...
+func TestTaskRunStatusWithSpire(t *testing.T) {
+	taskrunStatusTest(t, true)
+}
+
+func taskrunStatusTest(t *testing.T, spireEnabled bool) {
 	ctx := context.Background()
 	ctx, cancel := context.WithCancel(ctx)
 	defer cancel()
 	c, namespace := setup(ctx, t)
-	t.Parallel()
```
[the remark this answers is not in the document]
Comment (pxp928) [reply test/taskrun_test.go:143]: Because the `enable-spire` flag is changed when running with or without SPIRE. If the tests run in parallel, the config-map keeps changing while the non-spire and spire tests are running.

[the remark this answers is not in the document]
Comment (pxp928) [reply test/taskrun_test.go:143]: I wanted to keep both testing in place. One with spire (the original) and the addition of spire checks.

File: test/artifact_bucket_test.go [hunk 9]
```diff
@@ -258,26 +256,6 @@ spec:
 	}
 }
 
-// updateConfigMap updates the config map for specified @name with values. We can't use the one from knativetest because
-// it assumes that Data is already a non-nil map, and by default, it isn't!
-func updateConfigMap(ctx context.Context, client kubernetes.Interface, name string, configName string, values map[string]string) error {
```
[the remark this answers is not in the document]
Comment (pxp928) [reply test/artifact_bucket_test.go:263]: `updateConfigMap` was moved to init_test.go as its more generally to `enable-spire`

File: pkg/reconciler/taskrun/taskrun.go [hunk 10]
```diff
@@ -454,20 +461,40 @@ func (c *Reconciler) reconcile(ctx context.Context, tr *v1beta1.TaskRun, rtr *re
 	}
 
 	if podconvert.SidecarsReady(pod.Status) {
-		if err := podconvert.UpdateReady(ctx, c.KubeClientSet, *pod); err != nil {
-			return err
-		}
 		if err := c.metrics.RecordPodLatency(pod, tr); err != nil {
```
[the remark this answers is not in the document]
Comment (pxp928) [reply pkg/reconciler/taskrun/taskrun.go:464]: Yes I think you are right. I will move this after `podconvert.UpdateReady`

File: test/taskrun_test.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (priyawadhwa) [reply test/taskrun_test.go:143]: Could we run the non-spire tests first (in parallel) and then the spire tests (in parallel)? I'm not sure how much this will slow things down but it would probably be better to maintain parallelization if possible!

[the remark this answers is not in the document]
Comment (pxp928) [reply test/taskrun_test.go:143]: Sure I can change that.

[the remark this answers is not in the document]
Comment (pxp928) [reply test/taskrun_test.go:143]: @priyawadhwa any suggestions on how I can run the non-spire and spire tests separate? I want to avoid having multiple tests with the same code (with and without spire) and I don't think I have access to make another workflow (similar to the alpha feature flag) to sperate them out.

File: pkg/apis/config/feature_flags_test.go [hunk 11]
```diff
@@ -78,6 +79,16 @@ func TestNewFeatureFlagsFromConfigMap(t *testing.T) {
 			},
 			fileName: "feature-flags-bundles-and-custom-tasks",
 		},
+		{
+			expectedConfig: &config.FeatureFlags{
+				EnableAPIFields: "stable",
```
Comment (pritidesai) [inline pkg/apis/config/feature_flags_test.go:84]: Are we releasing this as a `stable` feature? or should this be behind `alpha` feature flag along with its own feature flag `enable-spire`?

File: pkg/apis/config/feature_flags_test.go [hunk 12]
```diff
@@ -95,6 +106,7 @@ func TestNewFeatureFlagsFromEmptyConfigMap(t *testing.T) {
 		RunningInEnvWithInjectedSidecars: true,
 		EnableAPIFields:                  "stable",
 		EmbeddedStatus:                   config.DefaultEmbeddedStatus,
+		EnableSpire:                      false,
```
Comment (pritidesai) [inline pkg/apis/config/feature_flags_test.go:109]: please use `config.DefaultEnableSpire` instead of hardcoding `false` here

File: pkg/apis/config/feature_flags_test.go (hunk 11 above)
Comment (pxp928) [reply pkg/apis/config/feature_flags_test.go:84]: Sure. I agree that it should be behind the alpha flag for the time being. I will change that.

File: test/e2e-tests.sh [hunk 13]
```diff
@@ -34,8 +34,12 @@ fi
 
 header "Setting up environment"
 
+install_spire
```
Comment (afrittoli) [inline test/e2e-tests.sh:37]: Since this is an alpha feature for now, I suppose we should only install SPIRE on the the job with alpha features enabled?

File: docs/spire.md [hunk 14]
```diff
@@ -0,0 +1,284 @@
+<!--
+---
+linkTitle: "SPIRE"
```
Comment (afrittoli) [inline docs/spire.md:3]: NIT: I think "TaskRun result attestation" could be a better title - SPIRE is the technology we decided to use, but it does not really describe the feature

File: docs/spire.md [hunk 15]
```diff
@@ -0,0 +1,284 @@
...
+weight: 1660
+---
+-->
+# SPIRE TaskRun result attestations
```
Comment (afrittoli) [inline docs/spire.md:7]: @geriom You may want to review this document as it will be a whole new page in the reference docs.

File: test/e2e-tests.sh (hunk 13 above)
Comment (pxp928) [reply test/e2e-tests.sh:37]: Yes I will incorporate that

File: docs/spire.md (hunk 14 above)
Comment (pxp928) [reply docs/spire.md:3]: Agreed. I will change it.

**Comment by pxp928** [comment]
> Thanks a lot for working on this feature @pxp928, and sorry about the delay in review. I'm going to release v0.37 today or tomorrow, so this won't make this release, but let's aim for getting this and possibly part2 in v0.38.
> 
> At a first glance this seems like a relatively sizeable change, so we may ask you to split this up, but I'll attempt to do an initial review first in any case.
> 
> I noticed the 0% unit test coverage on a few new modules:
> 
> ```
> pkg/spire/controller.go | Do not exist | 0.0% |  
> pkg/spire/entrypointer.go | Do not exist | 0.0%
> ```
> 
> Is this something you could look into? We have a few exported functions there, so we should provide coverage for them.
> 
> Thank you!!

The controller and entrypointer are covered via the mocked spire that we created. Plus those tests are covered under the integration tests. 

In hindsight this was a mistake on our part. We should have created smaller PRs to get this code incorporated. If that is what the maintainers require, I can work on getting smaller PRs with smaller additions.

**Comment by lumjjb** [comment]
For those 2 files, those are just wrappers around calling the spire implementation, we've refactored most of the code are in the rest of the files in the package so that they are re-used by the SPIRE mock for testing!

>I'd hate to delay your work on this further, but definitely smaller PRs are much easier to review with confidence

Thanks @afrittoli for the review! Do you have any recommendations on how to break this up? The individual components are a little hard to review without the context of how they are used. Was chatting with @pxp928 and think maybe we can schedule some time to help walk through the PR.

File: docs/spire.md [hunk 16]
```diff
@@ -0,0 +1,284 @@
...
+        #!/usr/bin/env bash
+        printf "%s" "hello" > "$(results.foo.path)"
+        printf "%s" "world" > "$(results.bar.path)"
+    results:
+    - name: foo
+    - name: bar
+```
+
+
+An example termination message would be:
```
Comment (geriom) [inline docs/spire.md:205]: Suggestion:

"The termination message is:"

File: docs/spire.md [hunk 17]
```diff
@@ -0,0 +1,284 @@
...
+        CERTIFICATE-----\nMIICCjCCAbCgAwIBAgIRALH94zAZZXdtPg97O5vG5M0wCgYIKoZIzj0EAwIwHjEL\nMAkGA1UEBhMCVVMxDzANBgNVBAoTBlNQSUZGRTAeFw0yMjAzMTQxNTUzNTlaFw0y\nMjAzMTQxNjU0MDlaMB0xCzAJBgNVBAYTAlVTMQ4wDAYDVQQKEwVTUElSRTBZMBMG\nByqGSM49AgEGCCqGSM49AwEHA0IABPLzFTDY0RDpjKb+eZCIWgUw9DViu8/pM8q7\nHMTKCzlyGqhaU80sASZfpkZvmi72w+gLszzwVI1ZNU5e7aCzbtSjgc8wgcwwDgYD\nVR0PAQH/BAQDAgOoMB0GA1UdJQQWMBQGCCsGAQUFBwMBBggrBgEFBQcDAjAMBgNV\nHRMBAf8EAjAAMB0GA1UdDgQWBBSsUvspy+/Dl24pA1f+JuNVJrjgmTAfBgNVHSME\nGDAWgBSOMyOHnyLLGxPSD9RRFL+Yhm/6qzBNBgNVHREERjBEhkJzcGlmZmU6Ly9l\neGFtcGxlLm9yZy9ucy9kZWZhdWx0L3Rhc2tydW4vbm9uLWZhbHNpZmlhYmxlLXBy\nb3ZlbmFuY2UwCgYIKoZIzj0EAwIDSAAwRQIhAM4/bPAH9dyhBEj3DbwtJKMyEI56\n4DVrP97ps9QYQb23AiBiXWrQkvRYl0h4CX0lveND2yfqLrGdVL405O5NzCcUrA==\n-----END
+        CERTIFICATE-----\n","type":1},{"key":"bar","value":"world","type":1},{"key":"bar.sig","value":"MEUCIQDOtg+aEP1FCr6/FsHX+bY1d5abSQn2kTiUMg4Uic2lVQIgTVF5bbT/O77VxESSMtQlpBreMyw2GmKX2hYJlaOEH1M=","type":1},{"key":"foo","value":"hello","type":1},{"key":"foo.sig","value":"MEQCIBr+k0i7SRSyb4h96vQE9hhxBZiZb/2PXQqReOKJDl/rAiBrjgSsalwOvN0zgQay0xQ7PRbm5YSmI8tvKseLR8Ryww==","type":1}]'
+```
+
+Parsed, the fields woud be:
```
Comment (geriom) [inline docs/spire.md:212]: Suggestion:

"Parsed, the fields are:"

File: docs/spire.md [hunk 18]
```diff
@@ -0,0 +1,284 @@
...
+              "-shell-image", "gcr.io/distroless/base@sha256:cfdc553400d41b47fd231b028403469811fcdbc0e69d66ea8030c5a0b5fbac2b",
+              # for script mode to work with windows we need a powershell image
+              # pinning to nanoserver tag as of July 15 2021
+              "-shell-image-win", "mcr.microsoft.com/powershell:nanoserver@sha256:b6d5ff841b78bdf2dfed7550000fd4f3437385b8fa686ec0f010be24777654d6",
+            ]
+    ```
+
+## Sample TaskRun attestation
+
+To demonstrate, we will use a simple task run that writes some results:
```
Comment (geriom) [inline docs/spire.md:182]: It's hard to parse the intention of this sentence. I suggest a more direct approach here

"The following example shows how this feature works:"

**Review by geriom — CHANGES_REQUESTED** [review]
Editorial review of the docs. Suggested a few small changes.

File: test/taskrun_test.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (afrittoli) [reply test/taskrun_test.go:143]: The problem with extra tests (workflows) is that each prow tests uses a GCP project from the Boskos pool, and we risk to starve CI jobs if we add too many e2e jobs. 
For changes that can be accessed via the API we have a single feature flag to avoid the explosion of test combinations. For backend features though we must have dedicated flags, but we don't have a good strategy in place to test them. I would assume spire enabled for now, but we should create a path to test the disable version too.
Would you mind filing an issue to track this problem in tektoncd/pluming (or tektoncd/pipeline)

**Comment by afrittoli** [comment]
> For those 2 files, those are just wrappers around calling the spire implementation, we've refactored most of the code are in the rest of the files in the package so that they are re-used by the SPIRE mock for testing!
> 
> > I'd hate to delay your work on this further, but definitely smaller PRs are much easier to review with confidence
> 
> Thanks @afrittoli for the review! Do you have any recommendations on how to break this up? The individual components are a little hard to review without the context of how they are used. Was chatting with @pxp928 and think maybe we can schedule some time to help walk through the PR.

Yeah, that's a fair point - but perhaps the spire package could be added by itself along with test coverage for it.
At least for me the highest barrier to review is the lack of knowledge about SPIRE, so having that in a dedicated PR might help.

File: test/taskrun_test.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (pxp928) [reply test/taskrun_test.go:143]: @afrittoli Just to understand this. You are saying, for now, just use the spire enabled by default for the e2e tests and not test the disabled one. I can create an issue to track this.

[the remark this answers is not in the document]
Comment (afrittoli) [reply test/taskrun_test.go:143]: Actually, rethinking about this, would it make sense to use the alpha job to test the spire integration and to run the spire disabled tests in the regular e2e job?

File: docs/spire.md [hunk 19]
```diff
@@ -0,0 +1,284 @@
...
+---
+linkTitle: "SPIRE"
+weight: 1660
+---
+-->
+# SPIRE TaskRun result attestations
+
+SPIRE TaskRun result attestations is currently an alpha experimental feature. 
+
+The SPIRE TaskRun result attestations feature provides the first part of non-falsifiable provenance to the build processes that run in the pipeline. They ensure that the results of the tekton pipeline executions originate from the build workloads themselves and that they have not been tampered. The second part of non-falsifiable provenance is to ensure that no third party interfered with the build process. Using SPIRE, the TaskRun status is monitored for any activity or change not preformed by the Tekton Pipeline Controller. If a unauthorized change is detected, it will invalidate the TaskRun.
```
[the remark this answers is not in the document]
Comment (pxp928) [reply docs/spire.md:11]: fixed

**Comment by pxp928** [comment]
> > For those 2 files, those are just wrappers around calling the spire implementation, we've refactored most of the code are in the rest of the files in the package so that they are re-used by the SPIRE mock for testing!
> > > I'd hate to delay your work on this further, but definitely smaller PRs are much easier to review with confidence
> > 
> > 
> > Thanks @afrittoli for the review! Do you have any recommendations on how to break this up? The individual components are a little hard to review without the context of how they are used. Was chatting with @pxp928 and think maybe we can schedule some time to help walk through the PR.
> 
> Yeah, that's a fair point - but perhaps the spire package could be added by itself along with test coverage for it. At least for me the highest barrier to review is the lack of knowledge about SPIRE, so having that in a dedicated PR might help.

@afrittoli I created another PR with jus the spire package -> https://github.com/tektoncd/pipeline/pull/5039

**Comment by lumjjb** [comment]
hmm with the updated milestone for 13 Sep for #4759 , it would be helpful if this can be reviewed soon as it will take some time to update #4759 with the changes here!

**Comment by dibyom** [comment]
@pxp928 is this ready for review or is there more work needed before that?

**Comment by lumjjb** [comment]
We are planning to break this up into smaller PRs to make it easier to
review! Should be ready around the 2nd week of November time frame!!
(Currently held up with kubecon prep for another project)

**Comment by pxp928** [comment]
> @pxp928 is this ready for review or is there more work needed before that?

Hey @dibyom, we are going to work to refactor the PRs to make them easier to review. With Kubecon around the corner, that might get done after.

**Comment by dibyom** [comment]
> Hey @dibyom, we are going to work to refactor the PRs to make them easier to review. With Kubecon around the corner, that might get done after.

Thanks for the update @pxp928 and @lumjjb ! I'll remove it from the v0.41 milestone then. Let me know when its ready and we'll add it back!

/cc @pritidesai @afrittoli @jagathprakash

**Comment by jerop** [comment]
doing a clean up of stale pull requests - feel free to reopen if needed

smaller pull requests: https://github.com/tektoncd/pipeline/pulls?q=is%3Apr+TEP-0089+

/close
